---
title: Journal Club
---

* TOC
{:toc}

{% if false %}
A brief description of the journal club activity is included in the [course rules](./Rules.html#journal-club), more details will be added later.
{% endif %}

## Submitting preferences

Please submit your preferences regarding these papers using [this form](https://forms.gle/K7BZJD5fA3wRmmXk8) until Wednesday Oct. 22, 2025, 22:00. In the form, include numbers of three papers from this list in the order of preference.
We will try to take your preferences into account, but we also need to balance group sizes and composition. 

* Groups will typically include students with different backgrounds.
* Preferences submitted earlier will be considered with a higher priority.

{% if false %}
* Also please indicate if you are willing to work in a group with English-speaking students (this means group discussions and final report in English)

Are you willing to take part in an English-speaking group?
Yes, I prefer English-speaking group
I can be in English-speaking group if needed
I strongly prefer Slovak-speaking group
{% endif %}


## List of papers


{% if false %}
## Groups
{% endif %}

* All links to papers should work in our University network (e.g.
    Eduroam or [proxy](https://uniba.sk/index.php?id=6828))

**Paper 1** Curry, K.D., Wang, Q., Nute, M.G., Tyshaieva, A., Reeves, E.,
Soriano, S., Wu, Q., Graeber, E., Finzer, P., Mendling, W. and Savidge,
T., 2022. Emu: species-level microbial community profiling of
full-length 16S rRNA Oxford Nanopore sequencing data. Nature methods,
19(7), pp.845-853. <https://doi.org/10.1038/s41592-022-01520-4>

**Paper 2** Edgar, R.C., Taylor, J., Lin, V., Altman, T., Barbera, P., Meleshko,
D., Lohr, D., Novakovsky, G., Buchfink, B., Al-Shayeb, B. and Banfield,
J.F., 2022. Petabase-scale sequence alignment catalyses viral discovery.
Nature, 602(7895), pp.142-147.
<https://doi.org/10.1038/s41586-021-04332-2>

**Paper 3.** Bradley, P., Den Bakker, H.C., Rocha, E.P., McVean, G. and Iqbal,
Z., 2019. Ultrafast search of all deposited bacterial and viral genomic
data. Nature biotechnology, 37(2), pp.152-159.
<https://doi.org/10.1038/s41587-018-0010-1> 

**Paper 4** Hauser, M., Steinegger, M. and SÃ¶ding, J., 2016. MMseqs software suite for fast and deep clustering and searching of large protein sequence sets. Bioinformatics, 32(9), pp.1323-1330. <https://doi.org/10.1093/bioinformatics/btw006>

**Paper 5** Wapinski, I., Pfeffer, A., Friedman, N. and Regev, A., 2007. Automatic genome-wide reconstruction of phylogenetic gene trees. Bioinformatics, 23(13), pp.i549-i558. <https://doi.org/10.1093/bioinformatics/btm193>

**Paper 6** Kress, A., Poch, O., Lecompte, O. and Thompson, J.D., 2023. Real or fake? Measuring the impact of protein annotation errors on estimates of domain gain and loss events. Frontiers in Bioinformatics, 3, p.1178926. <https://doi.org/10.3389/fbinf.2023.1178926>

**Paper 7** Gearing, L.J., Cumming, H.E., Chapman, R., Finkel, A.M., Woodhouse, I.B., Luu, K., Gould, J.A., Forster, S.C. and Hertzog, P.J., 2019. CiiiDER: A tool for predicting and analysing transcription factor binding sites. PloS one, 14(9), p.e0215495. <https://doi.org/10.1371/journal.pone.0215495>

**Paper 8** Huang, K.L., Scott, A.D., Zhou, D.C., Wang, L.B., Weerasinghe, A.,
Elmas, A., Liu, R., Wu, Y., Wendl, M.C., Wyczalkowski, M.A. and Baral,
J., 2021. Spatially interacting phosphorylation sites and mutations in
cancer. Nature communications, 12(1), p.2313.
<https://doi.org/10.1038/s41467-021-22481-w> 

**Paper 9** Gordon DE, Hiatt J, Bouhaddou M, Rezelj VV, Ulferts S, Braberg H,
Jureka AS, Obernier K, Guo JZ, Batra J, Kaake RM. Comparative
host-coronavirus protein interaction networks reveal pan-viral disease
mechanisms. Science. 2020 Dec 4;370(6521).
<https://doi.org/10.1126/science.abe9403>


## Group meeting 

  - Everyone should read the article before the first
    meeting.
  - The first meeting should take place no later than November 21. The
    meeting can be online (e.g. in MS Teams) or in person. Try to choose
    a time suitable for all members of the group.
  - At least 24 hours before the first group meeting, post its date,
    time and location to the [discussion forum in Moodle](https://moodle.uniba.sk/mod/forum/view.php?id=133956).
  - After the first meeting, one representative of the group will post a
    short summary to [Moodle](https://moodle.uniba.sk/mod/assign/view.php?id=133951) stating
    who attended the meeting, who apologized and who did not respond to
    the group and what was the outcome of the meeting - whether the
    group had problems with some parts of the article, whether it plans
    to meet again, etc. You do not need to report subsequent meetings.
  - If you need help from instructors, please contact us by email.
  - Failure to hold the first meeting according to these rules may
    affect the grade of the final report.

## Journal club report

  - In the report, describe the main methods and results of the article
    in your own words.
  - The report should be understandable for students of this course
    (both computer scientists and biologists), so explain the necessary
    terms that are not explained in the article and were not explained in class (do not repeat material from lectures).
  - You do not have to cover the entire content of the article in the
    report and, conversely, you can use other resources. Cite all sources you have used.
  - You should include the following (not necessarily in this order):
    - a brief overall description of the contents of the paper,
    - a detailed explanation of at least one computational method,
    - a detailed explanation of at least one biological result or evaluation of the method on data,
    - an explanation how the paper relates to the course material,
    - examples of other papers citing your paper and using or improving its results.
  - The recommended length is about 1-2 pages per person. However, the
    report should form one coherent text, so agree well on the outline
    and division of work and spend some time at the end on combining and
    editing individual contributions into a coherent report.
  - The report should list the members of the group who have actively
    participated in its work. Students who are not on this list will
    receive zero points for the report.


{% if false %}
## Unused papers

{% endif %}