{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H9I3dteeYGxB"
      },
      "source": [
        "## GEO2R in Colab\n",
        "\n",
        "Slightly modified version of the code produced by [GEO2R](https://www.ncbi.nlm.nih.gov/geo/geo2r/) for series [GSE161472](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE161472). Modifications done by Brona Brejova (marked as BB). Text commentary was also added by BB.\n",
        "\n",
        "* To set up R as the language  `Connect`, `Change runtime type`, `Runtime type` choose `R`.\n",
        "* To run all cells of the notebook: Main menu `Run all`.\n",
        "\n",
        "More information about DESeq2 package see <https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Installing libraries\n",
        "\n",
        "First we need to install DESeq2 library to be used for analysis and to load it in R."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "up_tU-lNj74W",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# BB: install libraries as ubuntu packages\n",
        "system(\"apt update\")\n",
        "system(\"apt install -y r-bioc-all r-bioc-geoquery r-bioc-deseq2\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "y0eIXrG-Urm5",
        "outputId": "a57f7d78-ceb4-4579-dfca-c8b80ee86f2f",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# Version info: R 4.2.2, Biobase 2.58.0, GEOquery 2.66.0, limma 3.54.0\n",
        "################################################################\n",
        "#   Differential expression analysis with DESeq2\n",
        "library(DESeq2)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Loading data from GEO database and performing basic transformations\n",
        "\n",
        "We selected data from article Luo, X., Yin, J., Dwyer, D., Yamawaki, T., Zhou, H., Ge, H., Han, C.Y., Shkumatov, A., Snyder, K., Ason, B. and Li, C.M., 2021. [Chamber-enriched gene expression profiles in failing human hearts with reduced ejection fraction. ](https://doi.org/10.1038/s41598-021-91214-2) Scientific reports, 11(1), p.11839.\n",
        "\n",
        "From all 84 experiments in this data set we selected samples coming from donors with non-failing hearts and separated them into samples from ventricles (19 samples) and atriums (17 samples). Ventricle is a larger chamber (komora) and atrium is a smaller chamber (predsie≈à). Samples from the left and right sides of the heart were combined."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "iM3bJQaXUuEf",
        "outputId": "22a89fd4-6282-470f-c767-7e0b8082c2cc",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# load counts table from GEO\n",
        "urld <- \"https://www.ncbi.nlm.nih.gov/geo/download/?format=file&type=rnaseq_counts\"\n",
        "path <- paste(urld, \"acc=GSE161472\", \"file=GSE161472_raw_counts_GRCh38.p13_NCBI.tsv.gz\", sep=\"&\");\n",
        "## BB: renames tbl to geo_tbl\n",
        "geo_tbl <- as.matrix(data.table::fread(path, header=T, colClasses=\"integer\"), rownames=\"GeneID\")\n",
        "\n",
        "# load gene annotations \n",
        "apath <- paste(urld, \"type=rnaseq_counts\", \"file=Human.GRCh38.p13.annot.tsv.gz\", sep=\"&\")\n",
        "annot <- data.table::fread(apath, header=T, quote=\"\", stringsAsFactors=F, data.table=F)\n",
        "rownames(annot) <- annot$GeneID\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Setting parameters and running DESeq2 analysis\n",
        "\n",
        "Table with results can be found in file table.tsv (see the folder icon in the left column menu). You can download it by right clicking on the file and selecting \"Download\"."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# sample selection\n",
        "gsms <- paste0(\"X000000000XXXXXXXXXXX000000000XXXXXXXXXXXX11111111\",\n",
        "        \"XXXXXXXXXXXX111111111XXXXXXXXXXXX\")\n",
        "sml <- strsplit(gsms, split=\"\")[[1]]\n",
        "\n",
        "# filter out excluded samples (marked as \"X\")\n",
        "sel <- which(sml != \"X\")\n",
        "sml <- sml[sel]\n",
        "# BB: renamed tbl below to geo_tbl\n",
        "tbl <- geo_tbl[ ,sel]\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NOZvmAMcghBm",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "# group membership for samples\n",
        "gs <- factor(sml)\n",
        "groups <- make.names(c(\"ventricle\", \"atrium\"))\n",
        "levels(gs) <- groups\n",
        "sample_info <- data.frame(Group = gs, row.names = colnames(tbl))\n",
        "\n",
        "# pre-filter low count genes\n",
        "# keep genes with at least N counts > 10, where N = size of smallest group\n",
        "keep <- rowSums( tbl >= 10 ) >= min(table(gs))\n",
        "tbl <- tbl[keep, ]\n",
        "\n",
        "ds <- DESeqDataSetFromMatrix(countData=tbl, colData=sample_info, design= ~Group)\n",
        "\n",
        "ds <- DESeq(ds, test=\"Wald\", sfType=\"poscount\")\n",
        "\n",
        "# extract results for top genes table\n",
        "r <- results (ds, contrast=c(\"Group\", groups[1], groups[2]), alpha=0.01, pAdjustMethod =\"fdr\")\n",
        "\n",
        "tT <- r[order(r$padj)[1:250],] \n",
        "tT <- merge(as.data.frame(tT), annot, by=0, sort=F)\n",
        "\n",
        "# BB: write the table to file instead of screen \n",
        "#tT <- subset(tT, select=c(\"GeneID\",\"padj\",\"pvalue\",\"lfcSE\",\"stat\",\"log2FoldChange\",\"baseMean\",\"Symbol\",\"Description\"))\n",
        "#write.table(tT, file=stdout(), row.names=F, sep=\"\\t\")\n",
        "write.table(tT, file=\"table.tsv\", row.names=F, sep=\"\\t\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## Visualization of results"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 894
        },
        "id": "aGEz0erGg61R",
        "outputId": "baf6415c-f635-4483-f89d-f9a4f0da9e52",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "plotDispEsts(ds, main=\"GSE161472 Dispersion Estimates\")\n",
        "\n",
        "# create histogram plot of p-values\n",
        "hist(r$padj, breaks=seq(0, 1, length = 21), col = \"grey\", border = \"white\", \n",
        "         xlab = \"\", ylab = \"\", main = \"GSE161472 Frequencies of padj-values\")\n",
        "\n",
        "# volcano plot\n",
        "old.pal <- palette(c(\"#00BFFF\", \"#FF3030\")) # low-hi colors\n",
        "par(mar=c(4,4,2,1), cex.main=1.5)\n",
        "plot(r$log2FoldChange, -log10(r$padj), main=paste(groups[1], \"vs\", groups[2]),\n",
        "     xlab=\"log2FC\", ylab=\"-log10(Padj)\", pch=20, cex=0.5)\n",
        "with(subset(r, padj<0.05 & abs(log2FoldChange) >= 0),\n",
        "     points(log2FoldChange, -log10(padj), pch=20, col=(sign(log2FoldChange) + 3)/2, cex=1))\n",
        "legend(\"bottomleft\", title=paste(\"Padj<\", 0.05, sep=\"\"), legend=c(\"down\", \"up\"), pch=20,col=1:2)\n",
        "\n",
        "# MD plot\n",
        "par(mar=c(4,4,2,1), cex.main=1.5)\n",
        "plot(log10(r$baseMean), r$log2FoldChange, main=paste(groups[1], \"vs\", groups[2]),\n",
        "     xlab=\"log10(mean of normalized counts)\", ylab=\"log2FoldChange\", pch=20, cex=0.5)\n",
        "with(subset(r, padj<0.05 & abs(log2FoldChange) >= 0),\n",
        "     points(log10(baseMean), log2FoldChange, pch=20, col=(sign(log2FoldChange) + 3)/2, cex=1))\n",
        "legend(\"bottomleft\", title=paste(\"Padj<\", 0.05, sep=\"\"), legend=c(\"down\", \"up\"), pch=20,col=1:2)\n",
        "abline(h=0)\n",
        "palette(old.pal) # restore palette\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "## General expression data visualization"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "CutGVVT7hDJP",
        "outputId": "646824be-278c-441d-b838-d353666ace14",
        "vscode": {
          "languageId": "r"
        }
      },
      "outputs": [],
      "source": [
        "################################################################\n",
        "#   General expression data visualization\n",
        "dat <- log10(counts(ds, normalized = T) + 1) # extract normalized counts\n",
        "\n",
        "# box-and-whisker plot\n",
        "lbl <- \"log10(raw counts + 1)\"\n",
        "ord <- order(gs)  # order samples by group\n",
        "palette(c(\"#1B9E77\", \"#7570B3\", \"#E7298A\", \"#E6AB02\", \"#D95F02\",\n",
        "          \"#66A61E\", \"#A6761D\", \"#B32424\", \"#B324B3\", \"#666666\"))\n",
        "par(mar=c(7,4,2,1))\n",
        "boxplot(dat[,ord], boxwex=0.6, notch=T, main=\"GSE161472\", ylab=\"lg(norm.counts)\", outline=F, las=2, col=gs[ord])\n",
        "legend(\"topleft\", groups, fill=palette(), bty=\"n\")\n",
        "\n",
        "# BB: umap omitted"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "R",
      "name": "ir"
    },
    "language_info": {
      "name": "R"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
